File: app/src/main/java/net/squanchy/search/view/SearchAdapter.java
package net.squanchy.search.view;

import android.app.Activity;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Collections;

import net.squanchy.R;
import net.squanchy.imageloader.ImageLoader;
import net.squanchy.schedule.view.EventItemView;
import net.squanchy.schedule.view.EventViewHolder;
import net.squanchy.search.SearchResult;

import static net.squanchy.imageloader.ImageLoaderComponentKt.imageLoaderComponent;

class SearchAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final ImageLoader imageLoader;
    private final Activity activity;

    @Nullable
    private SearchRecyclerView.OnSearchResultClickListener listener;

    private SearchResult.Success searchResult = new SearchResult.Success(Collections.emptyList(), Collections.emptyList());
    private ItemsAdapter itemsAdapter = new ItemsAdapter(searchResult);

    @IntDef({ViewTypeId.HEADER, ViewTypeId.SPEAKER, ViewTypeId.EVENT})
    @Retention(RetentionPolicy.SOURCE)
    @interface ViewTypeId {

        int HEADER = 1;
        int SPEAKER = 2;
        int EVENT = 3;
    }

    SearchAdapter(AppCompatActivity activity) {
        super();
        
        this.activity = activity;

        imageLoader = imageLoaderComponent(activity).imageLoader();
        setHasStableIds(true);
    }

    @Override
    @ViewTypeId
    public int getItemViewType(int position) {
        return itemsAdapter.viewTypeAtAbsolutePosition(position);
    }

    @Override
    public long getItemId(int position) {
        return itemsAdapter.itemIdAtAbsolutePosition(position);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, @ViewTypeId int viewType) {
        if (viewType == ViewTypeId.HEADER) {
            return new HeaderViewHolder(LayoutInflater.from(activity).inflate(R.layout.item_search_header, parent, false));
        } else if (viewType == ViewTypeId.SPEAKER) {
            View view = LayoutInflater.from(activity).inflate(R.layout.item_search_result_small, parent, false);
            return new SpeakerViewHolder(view);
        } else if (viewType == ViewTypeId.EVENT) {
            View view = LayoutInflater.from(activity).inflate(R.layout.item_schedule_event_talk, parent, false);
            return new EventViewHolder((EventItemView) view);
        } else {
            throw new IllegalArgumentException("Item type " + viewType + " not supported");
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        int viewType = itemsAdapter.viewTypeAtAbsolutePosition(position);

        if (viewType == ViewTypeId.SPEAKER) {
            ((SpeakerViewHolder) holder).updateWith(itemsAdapter.speakerAtAbsolutePosition(position), imageLoader, listener);
        } else if (viewType == ViewTypeId.HEADER) {
            ((HeaderViewHolder) holder).updateWith(itemsAdapter.headerTypeAtAbsolutePosition(position));
        } else if (viewType == ViewTypeId.EVENT) {
            ((EventViewHolder) holder).updateWith(
                    itemsAdapter.eventAtAbsolutePosition(position),
                    event1 -> {
                        if (listener != null) {
                            listener.onEventClicked(event1);
                        }
                        return null;
                    }
            );
        } else {
            throw new IllegalArgumentException("Item type " + viewType + " not supported");
        }
    }

    GridLayoutManager.SpanSizeLookup createSpanSizeLookup(int columnsCount) {
        return new GridSpanSizeLookup(itemsAdapter, columnsCount);
    }

    @Override
    public int getItemCount() {
        return itemsAdapter.totalItemsCount();
    }

    public void updateWith(SearchResult.Success searchResult, SearchRecyclerView.OnSearchResultClickListener listener) {
        this.searchResult = searchResult;
        this.itemsAdapter = new ItemsAdapter(searchResult);
        this.listener = listener;

        notifyDataSetChanged();
    }
}

File: app/src/main/java/net/squanchy/search/view/SearchAdapter.kt
package net.squanchy.search.view

import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.IntDef
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ListAdapter
import net.squanchy.R
import net.squanchy.imageloader.ImageLoader
import net.squanchy.imageloader.imageLoaderComponent
import net.squanchy.schedule.view.EventItemView
import net.squanchy.search.SearchListElement
import net.squanchy.search.SearchListElement.EventElement
import net.squanchy.search.SearchListElement.SpeakerElement
import net.squanchy.search.view.SearchItemViewHolder.AlgoliaLogoViewHolder
import net.squanchy.search.view.SearchItemViewHolder.HeaderViewHolder
import net.squanchy.search.view.SearchItemViewHolder.SearchEventViewHolder
import net.squanchy.search.view.SearchItemViewHolder.SpeakerViewHolder

internal class SearchAdapter(activity: AppCompatActivity) : ListAdapter<SearchListElement, SearchItemViewHolder>(DiffCallback) {

    private val imageLoader: ImageLoader
    private val activity: Activity

    private lateinit var listener: SearchRecyclerView.OnSearchResultClickListener

    init {
        this.activity = activity

        imageLoader = imageLoaderComponent(activity).imageLoader()
        setHasStableIds(true)
    }

    @ViewTypeId
    override fun getItemViewType(position: Int): Int {
        return when (getItem(position)) {
            is SearchListElement.EventHeader -> HEADER
            is SearchListElement.SpeakerHeader -> HEADER
            is EventElement -> EVENT
            is SpeakerElement -> SPEAKER
            is SearchListElement.AlgoliaLogo -> ALGOLIA_LOGO
        }
    }

    override fun getItemId(position: Int): Long {
        val item = getItem(position)
        return when (item) {
            is EventElement -> item.event.numericId
            is SpeakerElement -> item.speaker.numericId
            is SearchListElement.EventHeader -> ITEM_ID_EVENTS_HEADER
            is SearchListElement.SpeakerHeader -> ITEM_ID_SPEAKERS_HEADER
            is SearchListElement.AlgoliaLogo -> ITEM_ID_ALGOLIA_LOGO
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, @ViewTypeId viewType: Int): SearchItemViewHolder {
        when (viewType) {
            HEADER -> return HeaderViewHolder(LayoutInflater.from(activity).inflate(R.layout.item_search_header, parent, false))
            SPEAKER -> {
                val view = LayoutInflater.from(activity).inflate(R.layout.item_search_result_small, parent, false)
                return SpeakerViewHolder(view)
            }
            EVENT -> {
                val view = LayoutInflater.from(activity).inflate(R.layout.item_schedule_event_talk, parent, false)
                return SearchEventViewHolder(view as EventItemView)
            }
            ALGOLIA_LOGO -> {
                val view = LayoutInflater.from(activity).inflate(R.layout.item_search_algolia_logo, parent, false)
                return AlgoliaLogoViewHolder(view)
            }
            else -> throw IllegalArgumentException("Item type $viewType not supported")
        }
    }

    override fun onBindViewHolder(holder: SearchItemViewHolder, position: Int) {
        val item = getItem(position)
        when (item) {
            is SearchListElement.EventHeader -> (holder as HeaderViewHolder).updateWith(HeaderType.EVENTS)
            is SearchListElement.SpeakerHeader -> (holder as HeaderViewHolder).updateWith(HeaderType.SPEAKERS)
            is EventElement -> (holder as SearchEventViewHolder).updateWith(item.event) { listener.onEventClicked(it) }
            is SpeakerElement -> (holder as SpeakerViewHolder).updateWith(item.speaker, imageLoader, listener)
            is SearchListElement.AlgoliaLogo -> Unit // Nothing to do
        }
    }

    fun createSpanSizeLookup(columnsCount: Int): GridLayoutManager.SpanSizeLookup {
        return GridSpanSizeLookup(::getItem, ::isEmpty, columnsCount)
    }

    private fun isEmpty() = itemCount == 0

    fun updateWith(items: List<SearchListElement>, listener: SearchRecyclerView.OnSearchResultClickListener) {
        this.listener = listener
        super.submitList(items)
    }

    @Deprecated(
        message = "Use updateWith() instead",
        replaceWith = ReplaceWith("updateWith(items, onSearchResultClickListener)"),
        level = DeprecationLevel.ERROR
    )
    override fun submitList(items: MutableList<SearchListElement>?) {
        throw UnsupportedOperationException("Use updateWith() instead")
    }

    companion object {
        const val HEADER = 1
        const val SPEAKER = 2
        const val EVENT = 3
        const val ALGOLIA_LOGO = 4

        @IntDef(HEADER, SPEAKER, EVENT, ALGOLIA_LOGO)
        @Retention(AnnotationRetention.SOURCE)
        annotation class ViewTypeId

        // These values have been calculated manually using crc32, and they are never going to overlap
        // with the ones from speakers, events or tracks.
        private const val ITEM_ID_EVENTS_HEADER: Long = 3128461027
        private const val ITEM_ID_SPEAKERS_HEADER: Long = 1574748858
        private const val ITEM_ID_ALGOLIA_LOGO: Long = 3230264564
    }
}

private object DiffCallback : DiffUtil.ItemCallback<SearchListElement>() {
    override fun areItemsTheSame(oldItem: SearchListElement, newItem: SearchListElement): Boolean {
        return when {
            oldItem is EventElement && newItem is EventElement -> oldItem.event.numericId == newItem.event.numericId
            oldItem is SpeakerElement && newItem is SpeakerElement -> oldItem.speaker.numericId == newItem.speaker.numericId
            else -> oldItem == newItem
        }
    }

    override fun areContentsTheSame(oldItem: SearchListElement, newItem: SearchListElement): Boolean {
        return oldItem == newItem
    }
}
